provisioning/image_customization.go

```diff
@@ -242,12 +243,46 @@ func newImageCustomizationDeployment(info *ProvisioningInfo, ironicIPs []string,
 	}
 }
 
+func newImageCustomizationConfig(info *ProvisioningInfo, ironicIPs []string, inspectorIPs []string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        imageCustomizationConfigName,
+			Namespace:   info.Namespace,
+			Annotations: map[string]string{},
+			Labels: map[string]string{
+				"k8s-app":    metal3AppName,
+				cboLabelName: imageCustomizationService,
+			},
+		},
+		Type: corev1.SecretTypeOpaque,
+		StringData: map[string]string{
+			ironicAgentImage: info.Images.IronicAgent,
+			ironicBaseUrl:    getUrlFromIP(ironicIPs, baremetalIronicPort),
+			// TODO(dtantsur): when inspector is gone, we may be able to stop passing this URL
+			ironicInspectorBaseUrl: getUrlFromIP(inspectorIPs, baremetalIronicInspectorPort),
+			sshKeyEnvVar:           info.SSHKey,
+		},
+	}
+}
+
 func EnsureImageCustomizationDeployment(info *ProvisioningInfo) (updated bool, err error) {
 	ironicIPs, inspectorIPs, err := GetIronicIPs(info)
 	if err != nil {
 		return false, fmt.Errorf("unable to determine Ironic's IP to pass to the machine-image-customization-controller: %w", err)
 	}
 
+	configSecret := newImageCustomizationConfig(info, ironicIPs, inspectorIPs)
+	err = controllerutil.SetControllerReference(info.ProvConfig, configSecret, info.Scheme)
+	if err != nil {
+		err = fmt.Errorf("unable to set controllerReference on machine-image-customization config: %w", err)
+		return
+	}
+	_, _, err = resourceapply.ApplySecret(
+		context.Background(), info.Client.CoreV1(), info.EventRecorder, configSecret)
```

Comment:
Last I checked you'll need to use this wrapper function: https://github.com/openshift/cluster-baremetal-operator/blob/master/provisioning/baremetal_secrets.go#L51-L73 instead of calling straight into library-go, otherwise it ends up writing the Secret on every reconcile.

Reply:
Because it doesn't check if anything is changing before it writes it.

Reply:
It does though: https://github.com/openshift/library-go/blob/4109a9e7a43790f6bf68db64171b2257ae20d344/pkg/operator/resource/resourceapply/core.go#L423-L426. At least it intends to. Now, reading more the source, the problem could be the opposite: it assumes that the data never changes. I wish they had a usable documentation for these calls...

Reply:
If I'm reading it right, the wrapper is useless: since ApplySecret only compares metadata, it just exits. Could you double-check? Am I reading it wrong?

Reply:
This is the problematic part @zaneb: https://github.com/openshift/library-go/blob/4109a9e7a43790f6bf68db64171b2257ae20d344/pkg/operator/resource/resourceapply/core.go#L413. I suspect our only real option is to delete the secret...

Reply:
`.Data` is the content of the Secret, not the metadata, so it should be OK.

It looks like I misremembered the reason for adding our own wrapper, according to #127 it was so we could add custom logic like in https://github.com/openshift/cluster-baremetal-operator/blob/03222f4e09198bb7af9b7b8260fc2b8223f07553/provisioning/baremetal_secrets.go#L234-L240

I was definitely misled by https://github.com/openshift/cluster-baremetal-operator/blob/03222f4e09198bb7af9b7b8260fc2b8223f07553/provisioning/baremetal_secrets.go#L142-L175, which Get()s the Secret no less than 3 times - once before calling anything, again to pass to shallUpdateData() where it is ignored in favour of the first one, and finally in library-go which repeats the same comparison.

Reply:
#409

Reply:
> existingCopy is the thing that [gets written](https://github.com/openshift/library-go/blob/4109a9e7a43790f6bf68db64171b2257ae20d344/pkg/operator/resource/resourceapply/core.go#L439) at the end, that's why the data is copied into it.

Yeah, but only [if the metadata is different](https://github.com/openshift/library-go/blob/4109a9e7a43790f6bf68db64171b2257ae20d344/pkg/operator/resource/resourceapply/core.go#L423)?

Reply:
That's comparing the whole object, not just the metadata.
It fetches the object from the API, makes a copy, merges the new data that we're requiring into the copy, then compares the two to see if they're different.